### ISSUE TITLE
[1.20] iOS 3rd party FFI framework missing

### ISSUE THREAD

**Issue by vaind** [flutter/flutter#62853]
## Intro
I maintain a dart lib/flutter plugin making use of FFI (pub.dev/packages/objectbox) and have tried to test it with the current flutter beta channel. Turns out the `flutter build ios` build stopped including native "ObjectBox.framework" in the built test/example application. I've noticed the ios/Podfile is quite different now - were there any changes with regard to plugins/FFI, especially how the native libs are included? I'd prefer to address them before the next flutter release to prevent lib "downtime" if this isn't a flutter tool issue but an expected behaviour.

## Steps to Reproduce

1. Checkout https://github.com/vaind/obx-flutter-ios-build
2. Execute the following commands:
   ```
   flutter pub get
   flutter build ios --no-codesign --verbose
   ls -alh build/ios/iphoneos/Runner.app/Frameworks/
   ```
3. ObjectBox.framework is missing 

**Expected results:**
When built with a current stable flutter, the Frameworks directory includes `ObjectBox.framework`
```
$ flutter --version
Flutter 1.17.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 8af6b2f038 (5 weeks ago) • 2020-06-30 12:53:55 -0700
Engine • revision ee76268252
Tools • Dart 2.8.4
...
$ ls -alh build/ios/iphoneos/Runner.app/Frameworks/
...
drwxr-xr-x   4 objectbox  staff   128B Aug  4 13:39 ObjectBox.framework
...
```

**Actual results:** 
With the beta channel, it doesn't.
```
Flutter 1.20.0-7.4.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 916c3ac648 (3 days ago) • 2020-08-01 09:01:12 -0700
Engine • revision d6ee1499c2
Tools • Dart 2.9.0 (build 2.9.0-21.10.beta)
```
```
total 286960
drwxr-xr-x  12 objectbox  staff   384B Aug  4 14:33 .
drwxr-xr-x  27 objectbox  staff   864B Aug  4 14:33 ..
drwxr-xr-x   5 objectbox  staff   160B Aug  4 14:33 App.framework
drwxr-xr-x   6 objectbox  staff   192B Aug  4 14:33 Flutter.framework
-rw-r--r--   1 objectbox  staff    91M Aug  4 14:33 libswiftCore.dylib
-rw-r--r--   1 objectbox  staff   440K Aug  4 14:33 libswiftCoreFoundation.dylib
-rw-r--r--   1 objectbox  staff   1.9M Aug  4 14:33 libswiftCoreGraphics.dylib
-rw-r--r--   1 objectbox  staff   1.0M Aug  4 14:33 libswiftDarwin.dylib
-rw-r--r--   1 objectbox  staff   3.9M Aug  4 14:33 libswiftDispatch.dylib
-rw-r--r--   1 objectbox  staff    41M Aug  4 14:33 libswiftFoundation.dylib
-rw-r--r--   1 objectbox  staff   664K Aug  4 14:33 libswiftObjectiveC.dylib
drwxr-xr-x   4 objectbox  staff   128B Aug  4 14:33 path_provider.framework
```

## Details
I've noticed the difference in the "--verbose" run between stable and beta:
### Stable Flutter 1.17.5
```
[   +1 ms] CocoaPods' output:
           ↳
[        ]       Preparing
           
               Analyzing dependencies
           
               Inspecting targets to integrate
                 Using `ARCHS` setting to build architectures of target `Pods-Runner`: (``)
           
               Finding Podfile changes
                 - Flutter
                 - objectbox
                 - path_provider
                 - path_provider_linux
                 - path_provider_macos
           
               Fetching external sources
               -> Fetching podspec for `Flutter` from `Flutter`
               -> Fetching podspec for `objectbox` from `.symlinks/plugins/objectbox/ios`
               -> Fetching podspec for `path_provider` from `.symlinks/plugins/path_provider/ios`
               -> Fetching podspec for `path_provider_linux` from `.symlinks/plugins/path_provider_linux/ios`
               -> Fetching podspec for `path_provider_macos` from `.symlinks/plugins/path_provider_macos/ios`
           
               Resolving dependencies of `Podfile`
                 CDN: trunk Relative path: CocoaPods-version.yml exists! Returning local because checking is only
perfomed in
                 repo update
           
               Comparing resolved specification to the sandbox manifest
                 - Flutter
                 - objectbox
                 - path_provider
                 - path_provider_linux
                 - path_provider_macos
           
               Downloading dependencies
           
               -> Using Flutter (1.0.0)
           
               -> Using objectbox (0.0.1)
           
               -> Using path_provider (0.0.1)
           
               -> Using path_provider_linux (0.0.1)
           
               -> Using path_provider_macos (0.0.1)
                 - Running pre install hooks
```
### Beta Flutter 1.20.0-7.4.pre 
```
[   +2 ms] CocoaPods' output:
           ↳
[   +2 ms]       Preparing
           
               Analyzing dependencies
           
               Inspecting targets to integrate
                 Using `ARCHS` setting to build architectures of target `Pods-Runner`: (``)
           
               Fetching external sources
               -> Fetching podspec for `Flutter` from `Flutter`
               -> Fetching podspec for `path_provider` from `.symlinks/plugins/path_provider/ios`
           
               Resolving dependencies of `Podfile`
                 CDN: trunk Relative path: CocoaPods-version.yml exists! Returning local because checking is only
perfomed in
                 repo update
           
               Comparing resolved specification to the sandbox manifest
                 A Flutter
                 A path_provider
           
               Downloading dependencies
           
               -> Installing Flutter (1.0.0)
           
               -> Installing path_provider (0.0.1)
                 - Running pre install hooks
```

Any ideas?

**Comment by TahaTesser** [comment]
Hi @vaind 
Using your code sample and following the steps

### I get the following output and there is `ObjectBox.framework`

```
taha@Tahas-MacBook-Pro obx-flutter-ios-build % ls -alh build/ios/iphoneos/Runner.app/Frameworks/

total 286960
drwxr-xr-x  13 taha  staff   416B  5 Aug 16:37 .
drwxr-xr-x  27 taha  staff   864B  5 Aug 16:37 ..
drwxr-xr-x   5 taha  staff   160B  5 Aug 16:37 App.framework
drwxr-xr-x   6 taha  staff   192B  5 Aug 16:37 Flutter.framework
drwxr-xr-x   4 taha  staff   128B  5 Aug 16:37 ObjectBox.framework
-rw-r--r--   1 taha  staff    91M  5 Aug 16:37 libswiftCore.dylib
-rw-r--r--   1 taha  staff   440K  5 Aug 16:37 libswiftCoreFoundation.dylib
-rw-r--r--   1 taha  staff   1.9M  5 Aug 16:37 libswiftCoreGraphics.dylib
-rw-r--r--   1 taha  staff   1.0M  5 Aug 16:37 libswiftDarwin.dylib
-rw-r--r--   1 taha  staff   3.9M  5 Aug 16:37 libswiftDispatch.dylib
-rw-r--r--   1 taha  staff    41M  5 Aug 16:37 libswiftFoundation.dylib
-rw-r--r--   1 taha  staff   664K  5 Aug 16:37 libswiftObjectiveC.dylib
drwxr-xr-x   4 taha  staff   128B  5 Aug 16:37 path_provider.framework
```

<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel stable, v1.17.5, on Mac OS X 10.15.6 19G73, locale en-GB)
    • Flutter version 1.17.5 at /Users/taha/Code/flutter_stable
    • Framework revision 8af6b2f038 (5 weeks ago), 2020-06-30 12:53:55 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/taha/Code/sdk
    • Platform android-30, build-tools 30.0.1
    • ANDROID_HOME = /Users/taha/Code/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.1

[✓] Connected device (1 available)
    • SM M305F • 32003c30dc19668f • android-arm64 • Android 10 (API 29)

• No issues found!

```

</details>

**Comment by vaind** [comment]
Thanks for the replay @TahaTesser 

Yes, it works with the stable release (1.17) but doesn't with the upcoming 1.20 (e.g. beta channel). So the issue/question is about whether:
1. this is an intended change in flutter and how I (as a lib/plugin developer) can address it
2. this is an unintended change of the behaviour between flutter 1.17 and 1.20

**Comment by vaind** [comment]
Update: now that the new flutter version was released, this also reproducible on stable v1.20
```
Flutter 1.20.0 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 840c9205b3 (12 hours ago) • 2020-08-04 20:55:12 -0700
```

**Comment by TahaTesser** [comment]
Hi @vaind 
You're right, it's a regression 
```
```

**Comment by pcsosinski** [comment]
cc @cyanglaz

**Comment by jmagman** [comment]
pubspec:
```yaml
dependencies:
  flutter:
    sdk: flutter
...
  objectbox: ^0.6.4
```
.flutter-plugins-dependencies
```
{
   "info":"This is a generated file; do not edit or check into version control.",
   "plugins":{
      "ios":[
         {
            "name":"path_provider",
            "path":"/Users/m/.pub-cache/hosted/pub.dartlang.org/path_provider-1.6.11/",
            "dependencies":[

            ]
         }
      ],
      "android":[
         {
            "name":"path_provider",
            "path":"/Users/m/.pub-cache/hosted/pub.dartlang.org/path_provider-1.6.11/",
            "dependencies":[

            ]
         }
      ],
      "macos":[
         {
            "name":"path_provider_macos",
            "path":"/Users/m/.pub-cache/hosted/pub.dartlang.org/path_provider_macos-0.0.4+3/",
            "dependencies":[

            ]
         }
      ],
      "linux":[
         {
            "name":"path_provider_linux",
            "path":"/Users/m/.pub-cache/hosted/pub.dartlang.org/path_provider_linux-0.0.1+2/",
            "dependencies":[

            ]
         }
      ],
      "windows":[

      ],
      "web":[

      ]
   },
   "dependencyGraph":[
      {
         "name":"objectbox",
         "dependencies":[

         ]
      },
      {
         "name":"path_provider",
         "dependencies":[
            "path_provider_macos",
            "path_provider_linux"
         ]
      },
      {
         "name":"path_provider_linux",
         "dependencies":[

         ]
      },
      {
         "name":"path_provider_macos",
         "dependencies":[

         ]
      }
   ],
   "date_created":"2020-08-07 09:40:00.555565",
   "version":"1.20.1"
}
```
`objectbox` isn't in the `"ios` section so it's not being embedded.

**Comment by vaind** [comment]
FYI, looks like it's caused by the absence of plugin code (classes). We're working on a package update to get around that. 

Still a regression though, considering the previous flutter version worked fine .

**Comment by jmagman** [comment]
`objectbox` has had issues before https://github.com/flutter/flutter/issues/45778#issuecomment-580047017 which they point out in their README https://github.com/objectbox/objectbox-dart/tree/main/ios.  But I would still expect it to be in the `ios` section.

Their [pubspec](https://github.com/objectbox/objectbox-dart/blob/main/pubspec.yaml#L23-L26):
```yaml
flutter:
  plugin:
    android:
    ios:
```

**Comment by vaind** [comment]
> `objectbox` has had issues before [#45778 (comment)](https://github.com/flutter/flutter/issues/45778#issuecomment-580047017) which they point out in their README https://github.com/objectbox/objectbox-dart/tree/main/ios. But I would still expect it to be in the `ios` section.

Yep, that one was cocoapods not loading http source - that's been resolved (basically not using http source :) ) and has worked fine since then.

**Comment by jmagman** [comment]
For the "legacy" style pubspec ([without a `platforms:` key](https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms)) you're right, the `pluginClass` does seem to be required, even though the validator says this can be null.

https://github.com/flutter/flutter/blob/43c1b34cf5cb4517483a8e9e95277be6692c6f55/packages/flutter_tools/lib/src/plugins.dart#L283-L285

https://github.com/flutter/flutter/blob/43c1b34cf5cb4517483a8e9e95277be6692c6f55/packages/flutter_tools/lib/src/plugins.dart#L147-L149

**Comment by cyanglaz** [comment]
I wonder what was the intention of allowing the pluginClass to be null. Recently another change has added to master making this throw an error directly: https://github.com/flutter/flutter/commit/9da74f66ca2ef54054dc3b38377281c417efd5ae#diff-fb912d7008f6a340a5c4584a6ee1ae85

Running the example on master will show:
```
Invalid plugin specification objectbox.
Cannot find the `flutter.plugin.platforms` key in the `pubspec.yaml` file. An instruction to format the `pubspec.yaml` can be found
here: https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms
```

@csells Do you have any idea why this pubspec format was allowed?

CC @amirh

**Comment by jmagman** [comment]
On 1.17.x the ObjectBox framework is embedded in the app, but because it doesn't have a plugin class name it's not included in the generated `GeneratedPluginRegistrant` import list.  Which doesn't matter because the example app native iOS code doesn't actually need to import any symbols from the framework.  The tricksy bits are happening in dart:

https://github.com/objectbox/objectbox-dart/blob/5787d2b6b02ffdbeab850eae9b0bf39ec44a2e98/lib/src/bindings/bindings.dart#L172-L176

**Comment by vaind** [comment]
Because we wanted to restore the package functionality even before the flutter regression is fixed, I've ended up adding flutter-plugin native code (swift for ios) which resolved the issue and the native libs would be included (https://github.com/objectbox/objectbox-dart/pull/113).

However, there're some details you may be interested in for addressing this in flutter: `platforms` key is required now (with flutter v1.20.1) in the pubspec - without it, the dependency isn't recognized as a flutter plugin regardless of having the class specified; resulting in omission of the `objectbox.framework` from the built app.

In other words, this works:
```
flutter:
  plugin:
    platforms:
      android:
        package: io.objectbox.flutter.objectbox
        pluginClass: ObjectBoxFlutterPlugin
      ios:
        pluginClass: ObjectBoxFlutterPlugin
```
while this doesn't:
```
flutter:
  plugin:
    android:
      package: io.objectbox.flutter.objectbox
      pluginClass: ObjectBoxFlutterPlugin
    ios:
      pluginClass: ObjectBoxFlutterPlugin
```

**Comment by jmagman** [comment]
Thanks, I think @cyanglaz is already taking a look.

**Comment by cyanglaz** [comment]
@vaind Thank you for sharing the new workaround! This is probably the best workaround for 1.20.
It looks like after 1.20.0, we applied more strict rules on pubspec.yaml's plugin filed. The previous workaround for FFI plugins like `object_box` won't work anymore. 

We plan to work on supporting android/ios plugins without native(objc, java, swift, kotlin) `pluginClass` in some future release. And I have created an issue (https://github.com/flutter/flutter/issues/63545) to track that.

Meanwhile, I'm going to close this issue as there is no immediate plan to revive the support for the previous workaround in 1.20.
